[2.9.0] Feature Flag - external-rules

sunilarjun opened the pull request:
[SURE-8490]
Follow-up to https://github.com/rancher/rancher-docs/pull/1345.

Revising the "Feature Flag" reference page and adding a behavior change section regarding external `RoleTemplate` objects to the "Cluster and Project Roles" user guide page.

Fixes https://github.com/rancher/rancher-docs/issues/1357

martyav commented on docs/getting-started/installation-and-upgrade/installation-references/feature-flags.md:
```diff
@@ -19,7 +19,7 @@ Some feature flags require a restart of the Rancher container. Features that req
 
 - `continuous-delivery`: Allows Fleet GitOps to be disabled separately from Fleet. See [Continuous Delivery.](../../../how-to-guides/advanced-user-guides/enable-experimental-features/continuous-delivery.md) for more information.
-- `external-rules`: This flag is enabled by default. Only admin users can enable/disable the flag, and note that `escalate` permissions on `RoleTemplates` are required to create external `RoleTemplates` with `ExternalRules`. Restricted admin users can only enable the flag. If enabled, external `RoleTemplates` can be created only if the backing `ClusterRole` exists in the local cluster or the `Ext ...
+- `external-rules`: This flag is removed in this current and later Rancher versions as the flag always remains on. With its default enablement, external `RoleTemplates` can be created only if the backing `ClusterRole` exists in the local cluster or the `ExternalRules` is set. Note that `escalate` permissions on `RoleTemplates` are required to create external `RoleTemplates` with `ExternalRules`.  ...
```
```suggestion
- `external-rules`: In Rancher v2.9.0 and later, this flag is removed from the UI as the setting is permanently enabled. In previous Rancher versions, it is on by default. With this setting enabled, external `RoleTemplates` can be created only if the backing `ClusterRole` exists in the local cluster or the `ExternalRules` is set. Note that `escalate` permissions on `RoleTemplates` are required to create external `RoleTemplates` with `ExternalRules`. For context, the backing `ClusterRole` holds cluster rules and privileges, and shares the same `metadata.name` used in the `RoleTemplate` in your respective cluster referenced by the `ClusterRoleTemplateBinding/ProjectRoleTemplateBinding`.
```

martyav commented on versioned_docs/version-2.9/getting-started/installation-and-upgrade/installation-references/feature-flags.md:
```diff
@@ -19,6 +19,7 @@ Some feature flags require a restart of the Rancher container. Features that req
 The following is a list of feature flags available in Rancher. If you've upgraded from a previous Rancher version, you may see additional flags in the Rancher UI, such as `proxy` or `dashboard` (both [discontinued](/versioned_docs/version-2.5/reference-guides/installation-references/feature-flags.md)):
 
 - `continuous-delivery`: Allows Fleet GitOps to be disabled separately from Fleet. See [Continuous Delivery.](../../../how-to-guides/advanced-user-guides/enable-experimental-features/continuous-delivery.md) for more information.
+- `external-rules`: This flag is removed in this current and later Rancher versions as the flag always remains on. With its default enablement, external `RoleTemplates` can be created only if the backing `ClusterRole` exists in the local cluster or the `ExternalRules` is set. Note that `escalate` permissions on `RoleTemplates` are required to create external `RoleTemplates` with `ExternalRules`.  ...
```
1. If the flag isn't accessible to v2.9 users, I don't think we should add details about it to the v2.9 docs. 
2. But the information about how external `RoleTemplates` behave is always going to be relevant to v2.9.0, since the flag can no longer be disabled by the user. So maybe describe this behavior on a page related to cluster roles and role templates? https://ranchermanager.docs.rancher.com/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/manage-role-based-access-control-rbac/cluster-and-project-roles

sunilarjun replied:
The flag is no longer accessible to users so I ended up revising the "Feature Flag" page to point to a new section that describes the behavior change in the "Cluster and Project Roles" page mentioned, let me know what you think. Thanks Marty!

martyav reviewed (changes requested):
I left a suggestion about where we should describe the behavior controlled by this setting. Thanks!

martyav commented on docs/getting-started/installation-and-upgrade/installation-references/feature-flags.md:
```diff
@@ -19,7 +19,7 @@ Some feature flags require a restart of the Rancher container. Features that req
 
 - `continuous-delivery`: Allows Fleet GitOps to be disabled separately from Fleet. See [Continuous Delivery.](../../../how-to-guides/advanced-user-guides/enable-experimental-features/continuous-delivery.md) for more information.
-- `external-rules`: This flag is enabled by default. Only admin users can enable/disable the flag, and note that `escalate` permissions on `RoleTemplates` are required to create external `RoleTemplates` with `ExternalRules`. Restricted admin users can only enable the flag. If enabled, external `RoleTemplates` can be created only if the backing `ClusterRole` exists in the local cluster or the `Ext ...
+- `external-rules`: This flag affected [external `RoleTemplate` behavior](../../../how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/manage-role-based-access-control-rbac/cluster-and-project-roles.md#external-roletemplate-behavior) and is removed in v2.9.0 and later Rancher versions as the flag remains on by default.
```
```suggestion
- `external-rules`: This flag affected [external `RoleTemplate` behavior](../../../how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/manage-role-based-access-control-rbac/cluster-and-project-roles.md#external-roletemplate-behavior). It is removed in v2.9.0 and later as the behavior is enabled by default.
```

martyav commented on versioned_docs/version-2.9/getting-started/installation-and-upgrade/installation-references/feature-flags.md:
```diff
@@ -19,6 +19,7 @@ Some feature flags require a restart of the Rancher container. Features that req
 The following is a list of feature flags available in Rancher. If you've upgraded from a previous Rancher version, you may see additional flags in the Rancher UI, such as `proxy` or `dashboard` (both [discontinued](/versioned_docs/version-2.5/reference-guides/installation-references/feature-flags.md)):
 
 - `continuous-delivery`: Allows Fleet GitOps to be disabled separately from Fleet. See [Continuous Delivery.](../../../how-to-guides/advanced-user-guides/enable-experimental-features/continuous-delivery.md) for more information.
+- `external-rules`: This flag affected [external `RoleTemplate` behavior](../../../how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/manage-role-based-access-control-rbac/cluster-and-project-roles.md#external-roletemplate-behavior) and is removed in v2.9.0 and later Rancher versions as the flag always remains on.
```
```suggestion
`external-rules`: This flag affected [external `RoleTemplate` behavior](../../../how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/manage-role-based-access-control-rbac/cluster-and-project-roles.md#external-roletemplate-behavior). It is removed in Rancher v2.9.0 and later as the behavior is enabled by default.
```

sunilarjun replied:
Applied suggestion, thanks Marty!

martyav commented on docs/getting-started/installation-and-upgrade/installation-references/feature-flags.md:
```diff
@@ -19,7 +19,7 @@ Some feature flags require a restart of the Rancher container. Features that req
 
 - `continuous-delivery`: Allows Fleet GitOps to be disabled separately from Fleet. See [Continuous Delivery.](../../../how-to-guides/advanced-user-guides/enable-experimental-features/continuous-delivery.md) for more information.
-- `external-rules`: This flag is enabled by default. Only admin users can enable/disable the flag, and note that `escalate` permissions on `RoleTemplates` are required to create external `RoleTemplates` with `ExternalRules`. Restricted admin users can only enable the flag. If enabled, external `RoleTemplates` can be created only if the backing `ClusterRole` exists in the local cluster or the `Ext ...
+- `external-rules`: This flag affected [external `RoleTemplate` behavior](../../../how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/manage-role-based-access-control-rbac/cluster-and-project-roles.md#external-roletemplate-behavior) and is removed in v2.9.0 and later Rancher versions as the flag remains on by default.
```
```suggestion
`external-rules`: This flag affected [external `RoleTemplate` behavior](../../../how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/manage-role-based-access-control-rbac/cluster-and-project-roles.md#external-roletemplate-behavior). It is removed in Rancher v2.9.0 and later as the behavior is enabled by default.
```

sunilarjun replied:
Applied suggestion.

martyav reviewed (changes requested):
Thanks for the changes, I just suggested some slight rewording to make a sentence shorter.

raulcabello reviewed (dismissed):
LGTM

btat reviewed (commented):
Adding a LGTM for the wording, but the placement is a bit misleading since that list is prefaced as a list of flags available whereas this one is removed and not actually on the feature flags area of the UI. 

It might make more sense to move this entry into the subsequent table where the status column can be used to indicate this flag's (`Removed`) status and introduce another column for the context.

sunilarjun commented:
> It might make more sense to move this entry into the subsequent table where the status column can be used to indicate this flag's (`Removed`) status and introduce another column for the context.

Thanks for the review @btat! Just updated with your suggestion and moved the flag into the bottom table while adding a new row for additional information. Let me know what you think.

btat commented:
> > It might make more sense to move this entry into the subsequent table where the status column can be used to indicate this flag's (`Removed`) status and introduce another column for the context.
> 
> Thanks for the review @btat! Just updated with your suggestion and moved the flag into the bottom table while adding a new row for additional information. Let me know what you think.

Thanks @sunilarjun LGTM.

btat commented on docs/getting-started/installation-and-upgrade/installation-references/feature-flags.md:
```diff
@@ -34,14 +33,15 @@ The following is a list of feature flags available in Rancher. If you've upgrade
+| `legacy` | `false` for new installs, `true` for upgrades | GA | v2.6.0 | |
+| `rke1-custom-node-cleanup`| `true` | GA | v2.6.0 | |
+| `rke2` | `true` | Experimental | v2.6.0 | |
+| `token-hashing` | `false` for new installs, `true` for upgrades | GA | v2.6.0 | |
```
```suggestion
| `token-hashing` | `false` for new installs, `true` for upgrades | GA | v2.6.0 | |
| `ui-sql-cache` | `false` | Highly experimental | v2.9.0 |
```

Line 47 from this diff shouldn't be removed since it was just merged as part of https://github.com/rancher/rancher-docs/pull/1337.

btat commented on versioned_docs/version-2.9/getting-started/installation-and-upgrade/installation-references/feature-flags.md:
```diff
@@ -33,14 +33,14 @@ The following is a list of feature flags available in Rancher. If you've upgrade
+| `legacy` | `false` for new installs, `true` for upgrades | GA | v2.6.0 | |
+| `rke1-custom-node-cleanup`| `true` | GA | v2.6.0 | |
+| `rke2` | `true` | Experimental | v2.6.0 | |
+| `token-hashing` | `false` for new installs, `true` for upgrades | GA | v2.6.0 | |
```
```suggestion
| `token-hashing` | `false` for new installs, `true` for upgrades | GA | v2.6.0 | |
| `ui-sql-cache` | `false` | Highly experimental | v2.9.0 |
```
